### PR TITLE
Update design doc for PCE

### PR DESCRIPTION
**Pull request by rozdy**
Related to issue #2894

### CODE REVIEW

File: docs/design/path-computation.md [hunk 1]
```diff
@@ -28,12 +18,15 @@ The pce project has multiple unit tests that excercise most / all of the core lo
...
-2. Ensure you can run the acceptance tests:
-    1. make atdd tags="@PCE,@FPATH"
+1. Ensure you can run kilda-pce maven project, run the tests, and they all should pass.
 
-## Notea
+## Path computation strategies
 
-In addition to this sparse guide, there are more notes in PathComputationTests.java regarding how the acceptance is executed.  That should give you a good guide to how the code comes together.
+Strategy for each path computation should be provided explicitly. If flow
+create/update/reroute request doesn't has path strategy parameter then default
```
Comment (timofei-durakov) [inline docs/design/path-computation.md]: reroute should use flow specified param, please separate reroute from create/update

Comment (timofei-durakov) [reply docs/design/path-computation.md]: to make it clear by the moment of reroute some strategy will be already stored in flow itself, so we just reuse it

Comment (rozdy) [reply docs/design/path-computation.md]: Reworked doc.

**Comment by rozdy** [comment]
@surabujin 
> 
> > ```
> > Only links with **flow encapsulation type** support and sufficient bandwidth are used as available network resources.
> > ```
> 
> **flow encapsulation type** support => appropriate **flow encapsulation type**
> are => can be
> 
> 
> > ```
> > Manually assigned value for each link. If link is under maintenance and/or link is unstable then it's cost is increased by configurable weights.
> > ```
> 
> ... by preconfigured value.
> 

Fixed.

> _[docs/design/pce/design.md, line 55 at r1](https://reviewable.io/reviews/telstra/open-kilda/2895#-LsqidaR1GHEWPDIY5t8:-LsqidaR1GHEWPDIY5t9:b3an8cv) ([raw file](https://github.com/telstra/open-kilda/blob/2288554774d8ef4bc6122abd3fff36c74995d2d6/docs/design/pce/design.md#L55)):_
> 
> > ```gfm
> > * Latency
> > 
> > Automatically calculated value for each link. Configurable weight penalties is used when calculating link weight based on latency in a similar way as for cost.
> > ```
> 
> So... we are going to covert "latency"(time unit) value into "cost" or ... we are going to convert our "penalty" modifieds into "time" units and add to the measured latency value?

I guess separate modifiers for latency should be introduced. Updated doc accordingly.

File: docs/design/pce/design.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+Starting from source node we visit all neighbor nodes and saving path to all found nodes with lowest total weight. 
+
+Algorithm pseudocode:
+1. add source node to visit queue
+2. for each node in visit queue
+    1. if node is destination
+        1. if path weight is lower then current best 
+            1. save new path as current best path
+        2. skip further processing of the current node
```
Comment (timofei-durakov) [inline docs/design/pce/design.md]: nit: this means else? I'd make it explicit here and below

File: docs/design/pce/design.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+3. current best path is the best one
+4. if current best path is empty then no path is found
+
+### Weight function
+
+Weight function is defined by path computation strategy.
+For now two strategies are available:
+* Cost
+
+Manually assigned value for each link. If link is under maintenance and/or link is unstable then it's cost is increased by preconfigured value.
```
Comment (timofei-durakov) [inline docs/design/pce/design.md]: why not to Cost = Link Cost + Maint penalty + Unstable Penalty, where Main penalty = 0 if link isn't under maintenance...

File: docs/design/pce/design.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+* Latency
+
+Automatically calculated value for each link. Separate maintenance/unstable penalties is used when calculating link weight based on latency in a similar way as for cost.
```
Comment (timofei-durakov) [inline docs/design/pce/design.md]: this is very confusing, latency is something in ms/ns, maintenance/unstable penalties are some abstract numbers, how are you gonna sum them up?

File: docs/design/solutions/pce-weights-strategies/pce-weights-strategies.md [hunk 5]
```diff
@@ -1,23 +1,22 @@
...
+One path computer class parameterized with "get weight" strategy function is introduced instead of "computer class per cost strategy" design.
 
-## Proposed changes
-- Introduce PCE data model instead of reusing DAO: for storing graph weighs by features and more convenient logging
-- Decouple AvailableNetwork creation: it will use own building strategies.
-- Introduce weights computing strategies: like weight getters above data model, defines what weigh features will be used in path computation and in that proportions
+## Features
+- Introduced PCE data model instead of reusing DAO: for storing graph weighs by features and more convenient logging
+- Decoupled AvailableNetwork creation: it has own building strategies.
+- Introduced weight computing strategies: edge data model to long functions, defines what weigh features will be used in path computation and in what proportions
```
Comment (timofei-durakov) [inline docs/design/solutions/pce-weights-strategies/pce-weights-strategies.md]: I believe this should be covered in details

File: docs/design/pce/design.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+Criteria for best path computation is dependent on path computation strategy
+param. Available network resources are calculated with respect to encapsulation
+type param. Best paths are sorted in descending order of MinAvailableBandwidth. 
+
+## Algorithm
+
+Breadth-first search with path weight calculation and path length limitation 
+is used to calculate best path. 
+
+Starting from source node we visit all neighbor nodes and saving path to all found
```
[the remark this answers is not in the document]
Comment (rozdy) [reply docs/design/pce/design.md]: It's American English vs British English spelling :)

File: docs/design/pce/design.md (hunk 2 above)
Comment (rozdy) [reply docs/design/pce/design.md]: No, just next step in "then" block. Added explicit keywords.

File: docs/design/pce/design.md (hunk 3 above)
Comment (rozdy) [reply docs/design/pce/design.md]: Added example with a formula.

File: docs/design/pce/design.md (hunk 4 above)
Comment (rozdy) [reply docs/design/pce/design.md]: Correct, latency is measured in some time units. But when we use it to calculate abstract weight it becomes just a number (long). And then we will sum it with some other numbers (penalties). It's a big question on how to choose correct values and I guess it's out of scope of this design.

File: docs/design/solutions/pce-weights-strategies/pce-weights-strategies.md (hunk 5 above)
Comment (rozdy) [reply docs/design/solutions/pce-weights-strategies/pce-weights-strategies.md]: Added some details. Note: it's already done features. I just reworked this doc from "todo-style" to "done-style".

**Pull request merged**